packages/@atjson/editor/src/text-input.ts

```diff
@@ -0,0 +1,277 @@
+    });
+  } else if (node.nodeType === TEXT_NODE_TYPE) {
+    nodes.push(node as Text);
+  }
+
+  return nodes;
+}
+
+
+/**
+ * The keyboard mixin normalizes keyboard input across browsers.
+ * This is due to varying levels of support by browser vendors
+ * of different Web APIs. The Input Events API provides a fairly
+ * robust set of events that we can use to correctly detect input
+ * from people fluent in CJK languages (Chinese, Japanese, and Korean).
+ * These languages share a feature that _most_ input methods are
+ * done through a series of characters.
+ *
+ * For example, in Japanese, the following sequence of roman
+ * characters will result in the following set of text:
+ *
+ * | w | wa | wat | wata | watas | watash | watashi |
+ * | w | わ | わt | わた | わたs | わたsh | 私      |
+ *
+ * This results in text that does not map 1:1 to the keys that
+ * the person typed on their keyboard.
+ *
+ * This is the same series of events that we receive from
+ * autocorrect and predictive text keyboards.
+ *
+ * Our approach here is to do the best we can to get the most
+ * accurate set of events from the user's keyboard. We can only
+ * promise accuracy to the level of what is provided by the
+ * fidelity of the web API that's available for use in the browser.
+ */
+class TextInput extends events(HTMLElement) {
+  static events = {
+    'beforeinput': 'beforeinput',
+    'drag'(evt) {
+      this.lastDragEvent = evt;
+    },
+    'compositionend'(evt) {
+      let { start, end } = this.selection;
+      if (start === end) {
+        this.dispatchEvent(new CustomEvent('insertText', { detail: { position: start, text: evt.data } }));
+      } else {
+        this.dispatchEvent(new CustomEvent('replaceText', { detail: { start, end, text: evt.data } }));
+      }
+      this.composing = false;
+    },
+    'change text-selection'(evt) {
+      this.selection = evt.detail;
+    },
+    'clear text-selection'() {
+      this.selection = null;
+    }
+  };
+  private selection?: { start: number, end: number, collapsed: boolean } | null;
+
+  beforeinput(evt: InputEvent) {
+
+    if (evt.isComposing) return;
+
+    let ranges = evt.getTargetRanges();
+    let { start, end } = this.selection;
+
+    console.log('here is a thing', evt);
+
+    switch (evt.inputType) {
+    case 'insertText':
+      if (start === end) {
+        this.dispatchEvent(new CustomEvent('insertText', { detail: { position: start, text: evt.data } }));
+      } else {
+        this.dispatchEvent(new CustomEvent('replaceText', { detail: { start, end, text: evt.data } }));
+      }
+      break;
+
+    case 'insertLineBreak':
+      this.dispatchEvent(new CustomEvent('insertText', { detail: { position: start, text: '\u2028' } }));
+      this.dispatchEvent(new CustomEvent('addAnnotation', {
+        detail: { type: 'line-break', start, end: end + 1 }
+      }));
+      break;
+
+    case 'insertFromPaste':
+      var text = evt.dataTransfer.getData('text/plain');
```

Comment:
Urrrrgh. This is probably a different beast of a thing, where we have paste handlers?

So, for pasting HTML, we should probably have a registered paste handler that converts the HTML to an `HTMLSource`, then to the source that it's being edited in for offset. Same thing goes for Google Docs paste, Plain text paste, etc.

The other thing here is shortcode expansion. 🙃 So, when we get a Twitter URL, we expand it into a `tweet` annotation.